src/istore_agg.c

```diff
@@ -1,106 +1,276 @@
 #include "istore.h"
-#include "istore.h"
+
+#define SAMESIGN(a,b)   (((a) < 0) == ((b) < 0))
+#define INITSTATESIZE 30
+
+typedef struct{
+    size_t  size;
+    int     used;
+    BigIStorePair pairs[0];
+} ISAggState;
+
+static inline ISAggState *
+state_init(MemoryContext agg_context)
+{
+    ISAggState    *state;
+    state =  (ISAggState *) MemoryContextAllocZero(agg_context, sizeof(ISAggState) + INITSTATESIZE * sizeof(BigIStorePair));
+    state->size = INITSTATESIZE ;
+    return state;
+}
 
 /*
- * summarize an array of istores
+ * SUM(istore) aggregate funtion
  */
+PG_FUNCTION_INFO_V1(istore_sum_transfn);
 Datum
-istore_array_sum(Datum *data, int count, bool *nulls)
+istore_sum_transfn(PG_FUNCTION_ARGS)
 {
-    BigIStore       *out;
-    IStore          *istore;
-    AvlNode         *tree;
-    IStorePair      *payload;
-    BigIStorePairs  *pairs;
-    AvlNode         *position;
-    int              i,
-                     n,
-                     index;
 
+    MemoryContext  agg_context;
+    ISAggState    *state;
+    IStore        *istore;
+    IStorePair    *pairs2;
+    BigIStorePair *pairs1;
+    int            index1 = 0,
+                   index2 = 0;
 
-    tree = NULL;
-    n    = 0;
 
-    for (i = 0; i < count; ++i)
-    {
-        if (nulls[i])
-            continue;
+    if (!AggCheckCallContext(fcinfo, &agg_context))
+        elog(ERROR, "aggregate function called in non-aggregate context");
+
+    if (PG_ARGISNULL(1) && PG_ARGISNULL(0))
```

Comment:
Ah, we probably can't use strict definition as the state type and the arg type are not the same.